File: Algorithms/Dynamic Programming/DP on strings/wordBreak.py
def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        '''
        state: dp[i] -> can the word be segmented from this position
        transition: dp[i] |= dp[j] if s[i:j+1] in wordDict
        base case: dp[n] = True
        '''
        n = len(s)
        dp = [False] * (n + 1)
        dp[-1] = True

        for i in range(n-1, -1, -1):
            for word in wordDict:
                l = len(word)
                if s[i:i+l] == word:
                    dp[i] |= dp[i + l]

        return dp[0]

File: DSA-fasttrack-for-interviews/dp/src/word_break.rs
pub fn run(s: String, word_dict: Vec<String>) -> bool {
    let mut dp: Vec<bool> = vec![false; s.len() + 1];
    dp[0] = true;
    for i in 0..s.len() {
        if dp[i] {
            for word in &word_dict {
                if i + word.len() <= s.len() && &s[i..i + word.len()] == word {
                    dp[i + word.len()] = true;
                }
            }
        }
    }
    dp[s.len()]
}